### PR TITLE
adds the replace_html proc to tgwindow instances

### PR DESCRIPTION
**Pull request by azrael-x86**
## About The Pull Request

adds the ability to realtime inject HTML into TGWindow instances, just as you can create TGWindow instances with inline_html at initialization. 

the proc itself (`replace_html(inline_html as string)`) simply checks if the `inline_html` value actually has a value, and then sends a message with type `replace_html` and payload `inline_html`. `tgui.html` proceeds to do some wacky adjacent HTML and node shenanigans to entirely replace the inline HTML content. 

## Why It's Good For The Game

future usage with tgwindow instances

### CODE REVIEW

File: tgui/public/tgui.html [hunk 1]
```diff
@@ -487,6 +487,21 @@
   }
 };
 
+Byond.subscribeTo("replace_html", function(inline_html) {
+  let children = document.body.childNodes;
+
+  for (let i = 0; i < children.length; i++) {
```
Comment (stylemistake) [inline tgui/public/tgui.html:493]:
```suggestion
  var children = document.body.childNodes;

  for (var i = 0; i < children.length; i++) {
```

File: tgui/public/tgui.html [hunk 2]
```diff
@@ -487,6 +487,21 @@
   }
 };
 
+Byond.subscribeTo("replace_html", function(inline_html) {
```
Comment (stylemistake) [inline tgui/public/tgui.html:490]:
```suggestion
Byond.subscribeTo("replace_html", function(inline_html) {
```

**Comment by stylemistake** [comment]
A few things:

There is a small quirk in `Byond.subscribeTo()` which you're not aware of, is that there is a message queue, which collects sent messages while window is initializing, and at the moment `Byond.subscribeTo()` are reached, it flushes, and it stops queuing new messages. This is important, because TGUI takes some time to load because it fetches js bundles, and it has its own `Byond.subscribe()` function that loses all queued messages because it was called much later than `Byond.subscribeTo('replace_html')`.

It also doesn't need to be json encoded, so I think, this should bypass `send_message` altogether and instead use `<< output()` and `window.replaceHtml` callback (which is ok since it's an implementation detail).

Also you are removing comment nodes and replace them with new ones, which is a bit inefficient (you could be preserving them). And the new comment nodes that you reinsert, don't have an extra space around them, so your node matchers will fail.

File: tgui/public/tgui.html [hunk 3]
```diff
@@ -487,6 +487,21 @@
...
+    if (children[i].nodeValue == " tgui:inline-html-start ") {;
+      while (children[i].nodeValue != " tgui:inline-html-end ") {
+        children[i].remove();
+      }
+      children[i].remove();
+    }
+  }
+
+  document.body.insertAdjacentHTML("afterbegin", "<!-- tgui:inline-html-start -->" + inline_html + "<!-- tgui:inline-html-end -->");
+});
```
Comment (stylemistake) [inline tgui/public/tgui.html:503]:
```suggestion
Byond.subscribeTo("replace_html", function (inline_html) {
  var children = document.body.childNodes;

  for (var i = 0; i < children.length; i++) {
    if (children[i].nodeValue == " tgui:inline-html-start ") {
      while (children[i].nodeValue != " tgui:inline-html-end ") {
        children[i].remove();
      }
      children[i].remove();
    }
  }

  document.body.insertAdjacentHTML(
    "afterbegin",
    "<!-- tgui:inline-html-start -->"
      + inline_html
      + "<!-- tgui:inline-html-end -->"
  );
});
```

File: code/modules/tgui/tgui_window.dm [hunk 4]
```diff
@@ -311,6 +311,18 @@
...
+ * public
+ *
+ * Replaces the inline HTML content.
+ *
+ * required inline_html string HTML to inject
+ */
+/datum/tgui_window/proc/replace_html(inline_html)
+	if(!inline_html)
+		return
+	send_message("replace_html", inline_html)
```
Comment (stylemistake) [inline code/modules/tgui/tgui_window.dm:324]: I'd say you should do something like this, to allow completely replacing html with nothing.

```suggestion
/datum/tgui_window/proc/replace_html(inline_html = "")
	send_message("replace_html", inline_html)
```

File: code/modules/tgui/tgui_window.dm [hunk 5]
```diff
@@ -311,6 +311,16 @@
...
 
+/**
+ * public
+ *
+ * Replaces the inline HTML content.
+ *
+ * required inline_html string HTML to inject
+ */
+/datum/tgui_window/proc/replace_html(inline_html = "")
+	client << output(url_encode(inline_html), "statbrowser:replaceHtml")
```
Comment (stylemistake) [inline code/modules/tgui/tgui_window.dm:322]:
```suggestion
	client << output(url_encode(inline_html), "[window_id]:replaceHtml")
```

Comment (magatsuchi) [reply code/modules/tgui/tgui_window.dm:322]: WOOPS!

Comment (magatsuchi) [reply code/modules/tgui/tgui_window.dm]: you didnt see that

File: code/modules/tgui/tgui_window.dm [hunk 6]
```diff
@@ -311,6 +311,16 @@
...
 
+/**
+ * public
+ *
+ * Replaces the inline HTML content.
+ *
+ * required inline_html string HTML to inject
+ */
+/datum/tgui_window/proc/replace_html(inline_html = "")
+	client << output(url_encode(inline_html), "[window_id]:replaceHtml")
```
Comment (stylemistake) [inline code/modules/tgui/tgui_window.dm:322]:
```suggestion
	client << output(url_encode(inline_html), "[id].browser:replaceHtml")
```

**Review by stylemistake — APPROVED** [review]

**Pull request merged**